source/agora/consensus/data/UTXOSet.d

```diff
@@ -81,25 +126,89 @@ public class UTXOSet
 
         Params:
             tx = the transaction
+            height = Height of the block where UTXO will be unlocked
 
     ***************************************************************************/
 
-    public void updateUTXOCache (const ref Transaction tx) @safe
+    public void updateUTXOCache (const ref Transaction tx, ulong height) @safe
     {
-        foreach (const ref input; tx.inputs)
+        if (tx.type == TxType.Freeze)
         {
-            auto utxo_hash = getHash(input.previous, input.index);
-            this.utxo_db.remove(utxo_hash);
+            foreach (const ref input; tx.inputs)
+            {
+                auto utxo_hash = getHash(input.previous, input.index);
+                this.utxo_db.remove(utxo_hash);
+            }
+
+            ulong unlock_height = height + 1;
+
+            Hash tx_hash = tx.hashFull();
+            foreach (idx, output; tx.outputs)
+            {
+                auto utxo_hash = getHash(tx_hash, idx);
+                const UTXOSetValue value = UTXOSetValue(unlock_height, tx.type, output);
+                this.utxo_db[utxo_hash] = value;
+            }
         }
-
-        Hash tx_hash = tx.hashFull();
-        foreach (idx, output; tx.outputs)
+        else
         {
-            auto utxo_hash = getHash(tx_hash, idx);
-            this.utxo_db[utxo_hash] = output;
+            TxType tx_type = TxType.Payment;
+            UTXOSetValue utxo_value;
+            foreach (const ref input; tx.inputs)
+            {
+                if (this.getUTXOSetValue(input.previous, input.index, utxo_value))
+                {
+                    if (utxo_value.type == TxType.Freeze)
+                    {
+                        tx_type = TxType.Freeze;
+                        break;
+                    }
+                }
+            }
+
+            foreach (const ref input; tx.inputs)
+            {
+                auto utxo_hash = getHash(input.previous, input.index);
```

Comment:
This part seems to be the same for the freeze transactions too. Maybe it could be moved before the `if (tx.type == TxType.Freeze)` check so the code doesn't have to be duplicated.